Color interpolators

deciob opened the pull request:
This is still wip, but putting it here to better expose what I am doing.
There is an issue with the Interpolator.hcl method, but did not have enough time to dig into it, so at the moment I am not sure what is going on.

gampleman commented on src/Interpolation.elm:
```diff
@@ -286,6 +293,40 @@ hslImpl hueInt from to =
+                |> Debug.log "end -- this seems OK"
+    in
+    map4 fromHclPositional
+        (float start.hue end.hue)
```
Doesn't this one need to use the `hue` function?

```suggestion
        (hue start.hue end.hue)
```

gampleman replied:
Doesn't this one need to use the `hue` function?

```suggestion
        (hue start.hue end.hue)
```

See [here](https://github.com/d3/d3-interpolate/blob/master/src/hcl.js#L6) for reference.

deciob commented:
@gampleman I think this pull request is ready for review.

I have:
* Ported your work from https://github.com/avh4/elm-color/pull/23
* Added hcl & lab colour interpolations 
* Added an example page on colour interpolations

I still have one doubt on hclLong. It is returning rgb negative values instead of 0s. I have added a forcePositive function. The interpolation looks correct, but it is a bit of a hack. Not sure though where the originating issue is and if it is an issue at all.

gampleman commented on examples/ColorSpaceInterpolations.elm:
```diff
@@ -0,0 +1,231 @@
+    { fromColorValue = "#4682b4"
+    , toColorValue = "#ff0000"
+    , count = 10
+    }
```
I think making the start colours such that it clearly demonstrates some of the negative properties of RGB would be nice to "sell" this feature. For example, below you see how sRGB makes a "muddy" region in the middle.

<img width="818" alt="Screenshot 2020-03-09 at 15 05 24" src="https://user-images.githubusercontent.com/69144/76227588-7383aa80-6217-11ea-92dc-c8e182be2c18.png">

gampleman commented on examples/ColorSpaceInterpolations.elm:
```diff
@@ -0,0 +1,231 @@
+module ColorSpaceInterpolations exposing (main)
```
<img src="https://user-images.githubusercontent.com/69144/76228219-569ba700-6218-11ea-9dad-3d4d2f7479c7.png" />

I apologise for this not being documented anywhere clearly, but all examples are expected to fit into a 990 × 504 frame. This is the size of the iFrame they end up embedded in as well as the screenshots. This is what it will look like on the final site.

You can preview the example website here: https://5e665bbe58d21fbabc0f8a53--elm-visualization.netlify.com

gampleman commented on examples/ColorSpaceInterpolations.elm:
```diff
@@ -0,0 +1,231 @@
+import Html exposing (Html, div)
+import Html.Attributes exposing (class, for, href, id, selected, style, type_, value)
+import Html.Events exposing (onInput)
+import Interpolation exposing (Interpolator, hcl, hclLong, hsl, hslLong, lab, rgb)
```
I would prefer if (for didactical reasons) we could keep these functions prefixed. It makes it easier for readers to understand where these functions come from. (The type alias is fine)

gampleman commented on examples/ColorSpaceInterpolations.elm:
```diff
@@ -0,0 +1,231 @@
+                |> List.map (hclLong (hexToColor model.fromColorValue) (hexToColor model.toColorValue))
+                |> List.map Color.toRgba
+                |> List.map (\c -> ( c.red * 255 |> round, c.green * 255 |> round, c.blue * 255 |> round ))
+    in
```
I think this is unused?

deciob commented on src/Interpolation.elm:
```diff
@@ -1,8 +1,9 @@
+    , float, int, step, rgb, rgbWithGamma, hsl, hslLong, lab, hcl, hclLong
     , map, map2, map3, map4, map5, piecewise, tuple
     , inParallel, list, ListCombiner(..), combineParallel
+    , fromLab, toLab, fromHcl, toHcl
```
What do you think about moving them to the Color.Lab module?

gampleman replied:
I think that's fine, but at least as this PR goes the Color.Lab module is private, right?

deciob replied:
Yes it is.

gampleman replied:
I think moving those functions makes sense then.

deciob commented:
@gampleman I have squashed everything in a single commit. 
If you prefer keeping the full commit history, I have made a back up branch.
